### PR TITLE
Implements distributed lock with kubelock

### PR DESCRIPTION
**Pull request by TheoBrigitte**
Towards: https://github.com/giantswarm/giantswarm/issues/6691

In order to be able to run multiple aws-operator concurrently we need to make sure we do not allocate the same network to different clusters. For this we need a distributed lock during network allocation.

This PR wires the new [kubelock](https://github.com/giantswarm/kubelock/) library which act as the distributed lock here by writing an annotation on the `namespace/giantswarm` resource.

### CODE REVIEW

**Comment by TheoBrigitte** [comment]
This was tested and work as intended.

In order to test this I did the following:
* added an artificial sleep of 30 seconds right after the lock is acquired (so I had time to observe the second lock acquisition failing)
* ran 2 aws-operator (reconciling 2 different version 5.3.0, 5.2.1)
* create 2 cluster at the same time

Here are the results:

aws-operator-5.3.0 logs
```
D 19-08-29T12:45:16.766589+0 /apis/provider.giantswarm.io/v1alpha1/namespaces/default/awsconfigs/e4ysw ipamv29 acquiring lock for IPAM | aws-operator/service/network/allocator.go:43 | controller=aws-operator | event=update | loop=3 | version=148702214
D 19-08-29T12:45:16.775333+0 /apis/provider.giantswarm.io/v1alpha1/namespaces/default/awsconfigs/e4ysw ipamv29 acquired lock for IPAM | aws-operator/service/network/allocator.go:48 | controller=aws-operator | event=update | loop=3 | version=148702214

# lock is acquired now wait 30s
# removed lines about network allocation

D 19-08-29T12:45:46.98534+0 /apis/provider.giantswarm.io/v1alpha1/namespaces/default/awsconfigs/e4ysw ipamv29 persisted allocation candidate: "10.1.1.0/24" | aws-operator/service/network/allocator.go:93 | controller=aws-operator | event=update | loop=3 | version=148702214
D 19-08-29T12:45:46.985396+0 /apis/provider.giantswarm.io/v1alpha1/namespaces/default/awsconfigs/e4ysw ipamv29 releasing lock for IPAM | aws-operator/service/network/allocator.go:51 | controller=aws-operator | event=update | loop=3 | version=148702214
D 19-08-29T12:45:46.990648+0 /apis/provider.giantswarm.io/v1alpha1/namespaces/default/awsconfigs/e4ysw ipamv29 released lock for IPAM | aws-operator/service/network/allocator.go:56 | controller=aws-operator | event=update | loop=3 | version=148702214

# lock is released
```

aws-operator-5.2.1
```
D 19-08-29T12:45:20.729197+0 /apis/provider.giantswarm.io/v1alpha1/namespaces/default/awsconfigs/sj8y0 ipamv28patch1 acquiring lock for IPAM | aws-operator/service/network/allocator.go:43 | controller=aws-operator | event=update | loop=3 | version=148702263
W 19-08-29T12:45:20.732569+0 /apis/provider.giantswarm.io/v1alpha1/namespaces/default/awsconfigs/sj8y0 ipamv28patch1 retrying due to error | operatorkit/resource/wrapper/retryresource/resource_wrapper.go:74 | controller=aws-operator | event=update | loop=3 | underlyingResource=ipamv28patch1 | version=148702263
        /go/src/github.com/giantswarm/aws-operator/vendor/github.com/giantswarm/operatorkit/resource/wrapper/retryresource/resource_wrapper.go:67:
        /go/src/github.com/giantswarm/aws-operator/service/controller/legacy/v28patch1/resource/ipam/create.go:76:
        /go/src/github.com/giantswarm/aws-operator/service/network/allocator.go:46:
        /go/src/github.com/giantswarm/aws-operator/service/locker/distributed.go:92:
        /go/src/github.com/giantswarm/aws-operator/vendor/github.com/giantswarm/kubelock/namespaceable_lock.go:25:
        /go/src/github.com/giantswarm/aws-operator/vendor/github.com/giantswarm/kubelock/lock.go:36: lock `ipam` on `/api/v1/namespaces/giantswarm` owned by `aws-operator@2.0.0-0e25b29ac3735d8d170e6580c2d19dca55b90efa` already acquired at 2019-08-29T12:45:16Z with TTL 30s
        already exists error

# lock acquisition fails and keep retrying (removed retries)

D 19-08-29T12:46:08.250993+0 /apis/provider.giantswarm.io/v1alpha1/namespaces/default/awsconfigs/sj8y0 ipamv28patch1 acquiring lock for IPAM | aws-operator/service/network/allocator.go:43 | controller=aws-operator | event=update | loop=5 | version=148702268
D 19-08-29T12:46:38.259506+0 /apis/provider.giantswarm.io/v1alpha1/namespaces/default/awsconfigs/sj8y0 ipamv28patch1 acquired lock for IPAM | aws-operator/service/network/allocator.go:48 | controller=aws-operator | event=update | loop=5 | version=148702268

# lock is finally acquired after 30s
```

Here we can see the two different allocated networks for 2 clusters.
```
$ kubectl get awsconfig -ndefault -oyaml|yq -cr ".items[]|[.metadata.name,.status.cluster.network.cidr]"
["e4ysw","10.1.1.0/24"]
["sj8y0","10.1.11.0/24"]
```


Here is how the lock looks like
```
$ kubectl get ns giantswarm -oyaml
apiVersion: v1
kind: Namespace
metadata:
  annotations:
    kubelock.giantswarm.io/ipam: '{"owner":"aws-operator@2.0.0-0e25b29ac3735d8d170e6580c2d19dca55b90efa","createdAt":"2019-08-29T12:46:08.254392726Z","ttl":30000000000}'
  creationTimestamp: "2017-08-14T22:17:04Z"
  labels:
    name: giantswarm
  name: giantswarm
  resourceVersion: "148702411"
  selfLink: /api/v1/namespaces/giantswarm
  uid: 4d648588-813e-11e7-9908-025f39f4f185
spec:
  finalizers:
  - kubernetes
status:
  phase: Active
```

File: helm/aws-operator/templates/rbac.yaml [hunk 1]
```diff
@@ -63,8 +63,8 @@ rules:
       - drainerconfigs
     verbs:
       - "*"
-  # TODO drop namespaces when legacy bundles are gone and clusterapi bundle v29
-  # is the latest version.
+  # TODO drop create and delete permissions for namespaces when
+  # legacy bundles are gone and clusterapi bundle v29 is the latest version.
```
Comment (xh3b4sd) [inline helm/aws-operator/templates/rbac.yaml]: Let's keep the TODO comment in the other helm dir in sync.

File: service/locker/distributed.go [hunk 2]
```diff
@@ -0,0 +1,106 @@
+package locker
```
Comment (xh3b4sd) [inline service/locker/distributed.go]: Following the already established pattern the file should be called `kubelock_locker.go` and the implementation should be called `KubeLockLocker`.

File: service/locker/distributed.go [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+func NewDistributedLocker(config DistributedLockerConfig) (*DistributedLocker, error) {
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+	if config.RestConfig == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.RestConfig must not be empty", config)
+	}
+
+	var err error
+	var gvr schema.GroupVersionResource
```
Comment (xh3b4sd) [inline service/locker/distributed.go]: When this is only used in one place we could inline it.

File: service/locker/distributed.go [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+	err := d.kubelock.Lock(lockName).Acquire(ctx, lockNamespaceName, kubelock.AcquireOptions{
+		Owner: lockOwner,
+		TTL:   lockTTL,
+	})
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	return nil
+}
```
Comment (xh3b4sd) [inline service/locker/distributed.go]:
```suggestion
}

```

File: service/locker/distributed.go [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+		kubelock: k,
+	}
+
+	return d, nil
+}
+
+func (d DistributedLocker) Lock(ctx context.Context) error {
+	err := d.kubelock.Lock(lockName).Acquire(ctx, lockNamespaceName, kubelock.AcquireOptions{
+		Owner: lockOwner,
+		TTL:   lockTTL,
```
Comment (xh3b4sd) [inline service/locker/distributed.go]: How do we actually ensure that the action for which we want to lock is not timing out while we are operating?

Comment (tuommaki) [reply service/locker/distributed.go]: Do you mean how we ensure that lock doesn't timeout before the guarded operation finishes?

File: service/locker/distributed.go (hunk 2 above)
Comment (tuommaki) [reply service/locker/distributed.go]: +1 on this.

File: service/locker/distributed.go (hunk 5 above)
Comment (xh3b4sd) [reply service/locker/distributed.go]: Exactly. Like heartbeating or something.

Comment (xh3b4sd) [reply service/locker/distributed.go]: But I could imagine that our operation takes like 5 seconds and the TTL is 30 minutes. Then this should be safe enough for us for now. Just curious how we thought to address this eventually. /cc @kopiczko

File: service/locker/distributed.go (hunk 4 above)
Comment (TheoBrigitte) [reply service/locker/distributed.go]: What is this for ? There are no new lines at end of go files AFAIK.

File: service/locker/distributed.go (hunk 5 above)
Comment (TheoBrigitte) [reply service/locker/distributed.go]: Our operation takes under 1s and we have a 30s lock here, we are more than safe.

File: helm/aws-operator/templates/rbac.yaml (hunk 1 above)
Comment (TheoBrigitte) [reply helm/aws-operator/templates/rbac.yaml]: done

File: service/locker/distributed.go (hunk 2 above)
Comment (TheoBrigitte) [reply service/locker/distributed.go]: done

**Pull request merged**